crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,566 @@
+    VIRTIO_SND_PCM_FMT_U8, VIRTIO_SND_PCM_RATE_11025, VIRTIO_SND_PCM_RATE_16000,
+    VIRTIO_SND_PCM_RATE_176400, VIRTIO_SND_PCM_RATE_192000, VIRTIO_SND_PCM_RATE_22050,
+    VIRTIO_SND_PCM_RATE_32000, VIRTIO_SND_PCM_RATE_384000, VIRTIO_SND_PCM_RATE_44100,
+    VIRTIO_SND_PCM_RATE_48000, VIRTIO_SND_PCM_RATE_5512, VIRTIO_SND_PCM_RATE_64000,
+    VIRTIO_SND_PCM_RATE_8000, VIRTIO_SND_PCM_RATE_88200, VIRTIO_SND_PCM_RATE_96000,
+    VIRTIO_SND_S_BAD_MSG, VIRTIO_SND_S_NOT_SUPP,
+};
+use crate::Error;
+//equiv of PCMParams
+use crate::Stream;
+use std::{
+    cmp,
+    collections::HashMap,
+    convert::TryInto,
+    ops::Deref,
+    ptr,
+    ptr::NonNull,
+    sync::{Arc, RwLock},
+};
 
+use log::debug;
+use pipewire as pw;
+use pw::spa::param::ParamType;
+use pw::spa::pod::Pod;
+use std::ffi::CString;
+use std::mem::size_of;
+
+use pw::spa::sys::{
+    spa_audio_info_raw, SPA_AUDIO_CHANNEL_FC, SPA_AUDIO_CHANNEL_FL, SPA_AUDIO_CHANNEL_FR,
+    SPA_AUDIO_CHANNEL_LFE, SPA_AUDIO_CHANNEL_MONO, SPA_AUDIO_CHANNEL_RC, SPA_AUDIO_CHANNEL_RL,
+    SPA_AUDIO_CHANNEL_RR, SPA_AUDIO_CHANNEL_UNKNOWN, SPA_AUDIO_FORMAT_ALAW, SPA_AUDIO_FORMAT_F32,
+    SPA_AUDIO_FORMAT_F64, SPA_AUDIO_FORMAT_S16, SPA_AUDIO_FORMAT_S18_LE, SPA_AUDIO_FORMAT_S20,
+    SPA_AUDIO_FORMAT_S20_LE, SPA_AUDIO_FORMAT_S24, SPA_AUDIO_FORMAT_S24_LE, SPA_AUDIO_FORMAT_S32,
+    SPA_AUDIO_FORMAT_S8, SPA_AUDIO_FORMAT_U16, SPA_AUDIO_FORMAT_U18_LE, SPA_AUDIO_FORMAT_U20,
+    SPA_AUDIO_FORMAT_U20_LE, SPA_AUDIO_FORMAT_U24, SPA_AUDIO_FORMAT_U24_LE, SPA_AUDIO_FORMAT_U32,
+    SPA_AUDIO_FORMAT_U8, SPA_AUDIO_FORMAT_ULAW, SPA_AUDIO_FORMAT_UNKNOWN,
+};
+
+use pw::sys::{
+    pw_buffer, pw_loop, pw_thread_loop, pw_thread_loop_get_loop, pw_thread_loop_lock,
+    pw_thread_loop_new, pw_thread_loop_signal, pw_thread_loop_start, pw_thread_loop_unlock,
+    pw_thread_loop_wait, PW_ID_CORE,
+};
+use pw::{properties, spa, Context, Core, LoopRef};
+
+#[macro_export]
+macro_rules! properties_set {
+    ($properties:expr, $key:expr, $($args:tt)*) => {{
+        let key_cstr = CString::new($key).unwrap();
+        let format_str = format!($($args)*);
+        let format_cstr = CString::new(format_str).unwrap();
+        unsafe {
+            pw::sys::pw_properties_set(
+                $properties.as_ptr(),
+                key_cstr.as_ptr(),
+                format_cstr.as_ptr(),
+            )
+        }
+    }};
+}
+struct PwThreadLoop(NonNull<pw_thread_loop>);
+
+impl PwThreadLoop {
+    pub fn new(name: Option<&str>) -> Option<Self> {
+        let inner = unsafe {
+            pw_thread_loop_new(
+                name.map_or(ptr::null(), |p| p.as_ptr() as *const _),
+                std::ptr::null_mut(),
+            )
+        };
+        if inner.is_null() {
+            None
+        } else {
+            Some(Self(
+                NonNull::new(inner).expect("pw_thread_loop can't be null"),
+            ))
+        }
+    }
+
+    pub fn get_loop(&self) -> PwThreadLoopTheLoop {
```

Comment:
the name of this struct is pretty bad, i initially wrote that just to give you an idea of how to make the loop would work

Reply:
yes, I should probably change the struct name. thanks.